### DISCUSSION TITLE
xmake run cannot find shared libs

### DISCUSSION THREAD

**Discussion by qiuwei** [xmake-io/xmake#3690] [General]
The project compiles without any issue.

The project is organized as
```
/celere
xmake.lua
/gateway_xtp
     xmake.lua
```

the parent xmake.lua includes the subdirectiry `gateway_xtp` . the content for gateway_xtp is
```
target("xtp_md")
    set_kind("binary")

    add_deps("celere")

    add_includedirs("include", "external/2.2.33.5/include")
    add_linkdirs("external/2.2.33.5/lib")
    add_links("xtpquoteapi", "xtptraderapi")
    add_installfiles("external/2.2.33.5/lib/*", {prefixdir="lib"})
    add_syslinks("atomic", "pthread", "rt")

    add_headerfiles("include/(*/*.h)", "include/*.h", "include/(*/*.hpp)", "include/*.hpp", "include/(*/*/*.hpp)")
    add_files("src/**.cpp|xtp_td.cpp")
    add_includedirs("$(buildir)/$(plat)/$(arch)/$(mode)/")
```
`xmake run xtp_md` fails with
```
home/rnd/Projects/celere/build/linux/x86_64/release/xtp_md: error while loading shared libraries: libxtpquoteapi.so: cannot open shared object file: No such file or directory
``` 

I have used xmake run successfully in other projects using a similar setup, but could not find why it fails this time.
How can I check the env variables used by `xmake run`?  I tried add prints in `.local/share/xmake/modules/private/action/run/make_runenvs.lua` but could not see any output.

BTW, I am using `xmake v2.7.8+master.2ff74cf8e` on ubuntu 20.04

**Comment by waruqi** [comment]
you can use add_rpathdirs to add rpath.

**Reply by qiuwei** [reply]
I thought add_linkdirs would be enough for xmake run. 
If I understand it correctly, the rpath is supposed to work after the program is deployed. Apparently, the bin and lib directory layout would not be the same as for development.  So I am not sure rpath should be used in this case.

**Reply by waruqi** [reply]
> I thought add_linkdirs would be enough for xmake run.

make_runenvs just add `$PATH` on windows now, not `$LD_LIBRARY_PATH` on linux.

you can use `set_runenv` to set LD_LIBRARY_PATH. 

https://xmake.io/#/manual/project_target?id=targetset_runenv
